src/tableFormatter.ts

```diff
@@ -177,19 +215,34 @@ class MarkdownDocumentFormatter implements DocumentFormattingEditProvider {
                     jsLength -= cell.match(cjkRegex).length;
                 }
 
-                return this.alignText(cell, colAlign[i], jsLength);
+                if (delimiterRowNoPadding && rowNum === delimiterRowNum) {
+                    return this.alignText(cell, colAlign[i], jsLength, '');
+                }
+
+                return this.alignText(cell, colAlign[i], jsLength, cellPadding);
             });
-            return indentation + '| ' + cells.join(' | ') + ' |';
+
+            return indentation + '|' + cells.join('|') + '|';
```

Comment:
We can handle different cases of padding here (either " | " or "|" depending on `iRow`). And we don't have to change `alignText` at all.

Reply:
@yzhang-gh thanks for your feedback, but that wouldn't entirely work, I tried it that way.

The following case is then not handled right:

`md
| a |
| --- |
| x |`

Since the content of 'a' and 'x' are 1, but the minimal required width would be 3 (---, without padding). Thus that would lead to:

`md
| a   |
|---|
| x   |`

2 spaces would be added to the 'a' and 'x' because of their length = 1 and the cell must be 3. Then afterwards the padding is added (another 2 spaces) which makes the cell 5 characters wide, but the delimiter row only is 3 wide. 

Or the minimal width would stay 5. and that would lead to:

`md
| a   |
|-----|
| x   |`

Which is also not right since it can be 3 chars wide instead of 5.

Reply:
> (continued) So we don't have to introduce another `minimalWidth` and also `calculateNumberOfDelimiters` (just +2 if no padding).

For your `---` example,
- if padding (default), `return '-'.repeat(colWidth[i])`, which is `---`
- if no padding, `return '-'.repeat(colWidth[i] + 2)`, which is `-----`

https://github.com/yzhang-gh/vscode-markdown/pull/1110/files#diff-039c9fd55ddd5878168c5102ae45688a83a5fd0a21f8887acc017340608435a0L163-L165

Similarly, for `---:`,
- if padding (default), `return '-'.repeat(colWidth[i] - 1) + ':';`, which is `---:`
- if no padding, `return '-'.repeat(colWidth[i] - 1 + 2) + ':';`, which is `-----:`

Reply:
True, but what if the example is `---` and the value is `a` without padding.

Then the `a` column width = 3 (because of the minimal width), thus the align value would lead to `a  `, + cell join would be `| a   |` (which is 7 chars long)
Then the `---` column width = 3, thus the cell join would be `|---|` (which is 5 chars long. 

In this case you don't want the extra 2 `--` because that would lengthen the cell to much. And you would want to get the extra 2 spaces removed from the `a` value. That's what my code was doing.

Reply:
Emm... I don't quite understand this example.

I mean

0. the "content width" (without padding) is at least 3, up to the length of the longest cell (without padding)

Then we need to care about the padding in two places:
1. when we **normalize** the "delimiter-row", if padding, use the same "content width"; if no padding, use "content width"+2
   (in the `a` case, they should be `---` and `-----` respectively)
2. when we join the cells (of the "delimiter-row"), if padding, use " | "; if no padding, use "|"
   (in the `a` case, they should be `| --- |` and `|-----|` respectively)